lib/rules/prefer-regex-literals.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * @fileoverview Rule to disallow use of the `RegExp` constructor in favor of regular expression literals
+ * @author Milos Djermanovic
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const astUtils = require("./utils/ast-utils");
+const { CALL, CONSTRUCT, ReferenceTracker } = require("eslint-utils");
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Determines whether the given node is a string literal.
+ * @param {ASTNode} node Node to check.
+ * @returns {boolean} True if the node is a string literal.
+ */
+function isStringLiteral(node) {
+    return node.type === "Literal" && typeof node.value === "string";
+}
+
+/**
+ * Determines whether the given node is a template literal without expressions.
+ * @param {ASTNode} node Node to check.
+ * @returns {boolean} True if the node is a template literal without expressions.
+ */
+function isStaticTemplateLiteral(node) {
+    return node.type === "TemplateLiteral" && node.expressions.length === 0;
+}
+
+/**
+ * Determines whether the given node is a String.raw`` tagged template expression
+ * with a static template literal.
+ * @param {ASTNode} node Node to check.
+ * @returns {boolean} True if the node is String.raw`` with a static template.
+ */
+function isStringRawTaggedStaticTemplateLiteral(node) {
+    return node.type === "TaggedTemplateExpression" &&
+        node.tag.type === "MemberExpression" &&
+        node.tag.object.type === "Identifier" && node.tag.object.name === "String" &&
```

Comment:
Is there any way we can confirm that `String` was not overwritten, using similar tracking as we are using with `RegExp` elsewhere in this rule? (If it's too much of a pain to add, don't worry about it.)

Reply:
There is a way I did in https://github.com/eslint/eslint/pull/12252/files#diff-81df49cda58ea3f691055b0ce900e8e7R118.

Reply:
@mysticatea the test where the builtin is shadowed in the global scope is failing:

```js
"let String; new RegExp(String.raw`a`);",
```

probably because `variable.scope.type === "global"` is `true`.

Reply:
It's passing now with additional `variable.defs.length === 0` condition.

Reply:
Ah, in that case, we need to check `variable.defs.length ===0` as well. It means any declarations don't exist in the file.

Reply:
It's done now :)